src/main/scala/viper/gobra/util/ViperChopper.scala

```diff
@@ -0,0 +1,644 @@
+        val smallestProgramContainedInResult = smallestPrograms.forall(_.forall(containedInResult(_)))
+        assert(smallestProgramContainedInResult, "Chopper Error: Lost nodes during merging step.")
+
+        val containedInSmallest = Array.ofDim[Boolean](N)
+        for (program <- smallestPrograms; node <- program) { containedInSmallest(node) = true }
+        // checks all selected notes are present in the result
+        val selectedNodesContainedInSmallest = nodes.forall(containedInSmallest(_))
+        assert(selectedNodesContainedInSmallest, "Chopper Error: Not all selected nodes present in solution.")
+      }
+
+      mergedPrograms
+    }
+
+    /**
+      * Returns the set of smallest possible programs.
+      * @param N Number of nodes.
+      * @param nodes Nodes that must be included.
+      * @param edges Edges of the dependency graph.
+      * @return Set of smallest possible programs.
+      */
+    def smallestCut(N: Int, nodes: Vector[Int], edges: Array[mutable.SortedSet[Int]]): Vector[List[Int]] = {
+
+      /**
+        * Computes which of the nodes in `nodes` are dominating, i.e. not reached by other nodes in `nodes`,
+        * and then returns for each dominating node, the set of reachable nodes in a separate sorted list.
+        *
+        * This is done by applying dfs twice. The first run precomputes results that make the second run faster.
+        * The first run colors all nodes and computes the graph induced by the coloring,
+        * resulting in a smaller graph for the second run. All members of a color are sorted immediately to optimize
+        * intermediate results for the second run.
+        */
+
+      // Stores the color of nodes. A color of a node is only valid if the node was visited.
+      val coloring = Array.ofDim[Int](N)
+
+      // Stores whether a note was visited
+      val visited  = Array.ofDim[Boolean](N)
+
+      // Stores whether a color is not a root. Only valid if color was used.
+      val colorIsNotRoot  = Array.ofDim[Boolean](N)
+
+      def dfs1(start: Int, color: Int): (immutable.SortedSet[Int], List[Int]) = {
+        val stack = mutable.Stack[Int](start)
+        var members = immutable.SortedSet.empty[Int] // nodes of current color
+        var neighbors = List.empty[Int] // other colors reached from current color
+        val isAlreadyNeighbor = Array.ofDim[Boolean](color) // Assumes that `color` is max color used so far
+
+        while (stack.nonEmpty) {
+          val node = stack.pop()
+          if (!visited(node)) {
+            visited(node) = true
+            members += node
+            coloring(node) = color
+            stack.pushAll(edges(node))
+          } else {
+            val otherColor = coloring(node)
+            if (otherColor != color && !isAlreadyNeighbor(otherColor)) {
+              isAlreadyNeighbor(otherColor) = true
+              colorIsNotRoot(otherColor) = true
```

Comment:
Do I understand it correctly that in the case of a mutually recursive function, the coloring of the first visited function would be marked as non-root. It does however not matter which of the two functions is marked root or non-root because they will be placed in the same component. Mutually recursive methods do not even show up here because there's only a dependency on the other method's spec

Reply:
If there were two nodes that are mutually connected, then, yes, for them it does not matter which one is the root, since both will be members of the same color. However, as you noted, the mutual recursive methods do not show up here due to the indirection over the method spec.